File: test-projects/go-stdlib/_seeds/discount_test.go
package store

import "testing"

func pct(v int) DiscountCode   { return DiscountCode{Kind: "percent", Value: v} }
func fixedC(v int) DiscountCode { return DiscountCode{Kind: "fixed", Value: v} }

func TestApplyCodes(t *testing.T) {
	cases := []struct {
		name     string
		subtotal int
		codes    []DiscountCode
		want     int
	}{
		{"no codes", 2000, nil, 2000},
		{"percent", 2000, []DiscountCode{pct(10)}, 1800},
		{"fixed", 2000, []DiscountCode{fixedC(500)}, 1500},
		{"stack pct then fixed", 2000, []DiscountCode{pct(10), fixedC(500)}, 1300},
		{"stack fixed then pct", 2000, []DiscountCode{fixedC(500), pct(10)}, 1350},
		{"floor at zero", 1000, []DiscountCode{fixedC(1500)}, 0},
		{"percent clamp", 2000, []DiscountCode{pct(150)}, 0},
		{"rounding", 1299, []DiscountCode{pct(33)}, 870},
	}
	for _, c := range cases {
		if got := ApplyCodes(c.subtotal, c.codes); got != c.want {
			t.Errorf("%s: ApplyCodes(%d) = %d, want %d", c.name, c.subtotal, got, c.want)
		}
	}
}

func TestCartTotalAfterCodes(t *testing.T) {
	cart := []LineItem{{Qty: 2, PriceCents: 500}, {Qty: 1, PriceCents: 1299}}
	if got := CartTotalAfterCodes(cart, []DiscountCode{pct(10)}); got != 2069 {
		t.Errorf("CartTotalAfterCodes = %d, want 2069", got)
	}
}

File: test-projects/pnpm-vitest-monorepo/_seeds/discount.test.ts
// Seed (Full[Moderate]): the spec for stacking discount codes. The task's `seed` copies this to
// packages/core/src/discount.test.ts before the commit. Mid-level: the rules (running total, order
// dependence, floor at 0, percent clamp, rounding) + wiring into the cart take real care.
import { expect, test } from "vitest";
import { applyCodes, cartTotalAfterCodes, type DiscountCode } from "./discount.ts";

const pct = (value: number): DiscountCode => ({ kind: "percent", value });
const fixed = (value: number): DiscountCode => ({ kind: "fixed", value });

test("no codes returns the subtotal unchanged", () => {
  expect(applyCodes(2000, [])).toBe(2000);
});

test("a percent code takes that percent off", () => {
  expect(applyCodes(2000, [pct(10)])).toBe(1800);
});

test("a fixed code subtracts cents", () => {
  expect(applyCodes(2000, [fixed(500)])).toBe(1500);
});

test("codes stack IN ORDER (percent-then-fixed differs from fixed-then-percent)", () => {
  expect(applyCodes(2000, [pct(10), fixed(500)])).toBe(1300); // 2000 → 1800 → 1300
  expect(applyCodes(2000, [fixed(500), pct(10)])).toBe(1350); // 2000 → 1500 → 1350
});

test("the total never goes below zero", () => {
  expect(applyCodes(1000, [fixed(1500)])).toBe(0);
  expect(applyCodes(1000, [fixed(1500), pct(50)])).toBe(0);
});

test("a percent over 100 is clamped to 100% (free, never negative)", () => {
  expect(applyCodes(2000, [pct(150)])).toBe(0);
});

test("percent discounts round to the nearest cent", () => {
  expect(applyCodes(1299, [pct(33)])).toBe(870); // 1299 - 428.67 = 870.33 → 870
});

test("cartTotalAfterCodes applies codes to the cart subtotal", () => {
  const cart = [
    { qty: 2, priceCents: 500 },
    { qty: 1, priceCents: 1299 },
  ]; // subtotal 2299
  expect(cartTotalAfterCodes(cart, [pct(10)])).toBe(2069); // 2299 * 0.9 = 2069.1 → 2069
});
